### PR TITLE
feat; Tabs Capturing

### PR DESCRIPTION
**Pull request by saadberry**
## Changes
- Created a new directory `~/.codesync/.tabs`
- Updated assertions to ensure `.tabs` was being created
- Added a listener for tab changes.
- The pre-requisites for listening to tab changes are:
   1. Must be a valid user.
   2. Repo must be connected.
- Tab changes are recorded in two cases: 
   1.  When the IDE is opened.
   2.  When a user opens/closes tabs.
- Structuring data in the format:
```
repo_id: 1234 
created_at: <timestamp>
source: vscode
file_name: 123232134343.yml
tabs: 
- file_id: <file_id> path: <path>     
- file_id: <file_id_1> path: <path_1>
```
- Dumping tabs data in YML file as `<timestamp>.yml`
- Discarding files that are not from current repo
- Updated logic to prevent YML file being created if no tabs found
Setting `isActiveTab` to default value if  `!isTabEvent || !this.shouldProceed`
- Capturing initial state of tabs for when repo is connected
- Added further assertions for WebSocket Events related to tabs, specifically testing the status codes received from server. Namely:
-- Status 400 recieved.
-- Status 403 recieved.
-- Status 402 recieved.
- Fixed some failing tests.

### CODE REVIEW

**Review by Basit-qc — APPROVED** [review]

**Comment by Basit-qc** [comment]
Looks good @saadberry, one note; we should create PRs against `develop` branch. 
`develop` to `master` should only be for deployment purpose.

File: src/codesyncd/handlers/tabs_handler.ts [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+import { getRandomIndex, getTabsBeingProcessed } from "../utils";
+import { removeFile } from "../../utils/file_utils";
+import { TAB_FILES_PER_ITERATION, TAB_SIZE_LIMIT } from "../../constants";
+import { CodeSyncLogger } from "../../logger";
+import { TabValidator } from "../validators/tab_validator";
+import { TabHandler } from "./tab_handler";
+import { ConfigUtils } from "../../utils/config_utils";
+
+export class TabsHandler {
+    // @ts-ignore
```
Comment (Basit-qc) [inline src/codesyncd/handlers/tabs_handler.ts:17]: Why do we have @ts-ignore here? @saadberry

File: src/codesyncd/websocket/socket_events.ts [hunk 2]
```diff
@@ -97,11 +107,34 @@ export class SocketEvents {
...
+        // Send tabs
+        let validTabs: ITabYML[] = [];
+        errorCount = 0;   
+        const tabsHandler = new TabsHandler(this.repoTabs, this.accessToken);
+        const validTabsData = await tabsHandler.run();
+        if (!validTabsData) return;
+        validTabs = validTabs.concat(validTabsData);
+        // Keep track of tabs in State
+        const currentTabs = new Set(
+            validTabs.flatMap(validTab => validTab.tabs.map(tab => tab.path))
```
Comment (Basit-qc) [inline src/codesyncd/websocket/socket_events.ts:120]: @saadberry Why do we need flatMap here?

File: src/codesyncd/handlers/buffer_handler.ts [hunk 3]
```diff
@@ -223,20 +231,36 @@ export class bufferHandler {
...
+					tabsData = tabsHandler.getTabsData(tabYMLFiles.files);
+				}
+
+				if (diffs.files.length > 0) {
+					CodeSyncLogger.debug(`Processing ${diffs.files.length}/${diffs.count} diffs, uuid=${this.instanceUUID}`);
+					repoDiffs = this.groupRepoDiffs(diffs.files);
+				}
+			}
+
+			if (!tabYMLFiles.files.length && !diffs.files.length) return CodeSyncState.set(CODESYNC_STATES.BUFFER_HANDLER_RUNNING, false);;
```
Comment (Basit-qc) [inline src/codesyncd/handlers/buffer_handler.ts:259]: This can be moved to line 246 (Fail Fast)

File: src/codesyncd/handlers/tab_handler.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+import path from "path";
+
+import { ITabYML } from "../../interface";
+import {generateSettings} from "../../settings";
+import { removeTabFile } from "../../utils/tab_utils";
+import { isRelativePath } from "../utils";
+
+export class TabHandler {
+	
+	constructor() {}
```
Comment (Basit-qc) [inline src/codesyncd/handlers/tab_handler.ts:11]: If the constructor is empty, should we consider defining `static` methods?

File: src/codesyncd/handlers/tabs_handler.ts [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+export class TabsHandler {
+    tabYmlFiles : ITabYML[];
+    accessToken: string;
+
+    settings: any;
+    configJSON: any;
+    configRepo: any;
+
+    constructor(repoTab: ITabYML[] | null = null, accessToken: string | null = null) {
```
Comment (Basit-qc) [inline src/codesyncd/handlers/tabs_handler.ts:23]: If both params can be null, what's the point for constructor? Maybe should consider defining Static Method that does not need a constructor to be called

File: src/codesyncd/handlers/tabs_handler.ts [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        // @ts-ignore
```
Comment (Basit-qc) [inline src/codesyncd/handlers/tabs_handler.ts:24]: Should remove 
```// @ts-ignore```
from here and the line just below

File: src/codesyncd/validators/tab_validator.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+	configJSON: any;
+	constructor() {
+		this.settings = generateSettings();
+        this.configJSON = readYML(this.settings.CONFIG_PATH);
+
+	}
+
+	validateRepo( repoPath: string ) {
+		const configRepo = this.configJSON.repos[repoPath];
+		if (!configRepo) {
```
Comment (Basit-qc) [inline src/codesyncd/validators/tab_validator.ts:19]: we can remove parantheises

File: src/codesyncd/websocket/socket_events.ts [hunk 8]
```diff
@@ -95,13 +105,37 @@ export class SocketEvents {
             } else {
                 setDiffsBeingProcessed(currentDiffs);
             }
+
             this.connection.send(JSON.stringify({"diffs": validDiffs}));
         }
+
+        // Send tabs
+        let validTabs: ITabYML[] = [];
```
Comment (Basit-qc) [inline src/codesyncd/websocket/socket_events.ts:113]: should it be `const`?

File: src/codesyncd/websocket/socket_events.ts [hunk 9]
```diff
@@ -115,6 +149,22 @@ export class SocketEvents {
         setDiffsBeingProcessed(diffsBeingProcessed);
     }
 
+    async onTabProcessed(tabFileName: string) {
+        const tabHandler = new TabHandler();
```
Comment (Basit-qc) [inline src/codesyncd/websocket/socket_events.ts:153]: Should add 
```if (!this.canSendSocketData) return;```
Keeping in view `onSyncSuccess`

File: src/init/utils.ts [hunk 10]
```diff
@@ -13,12 +13,13 @@ import { checkServerDown } from '../utils/api_utils';
 import { IFileToUpload } from '../interface';
 import { uploadRepoToServer } from '../utils/upload_utils';
 import { CONNECTION_ERROR_MESSAGE, VSCODE, NOTIFICATION, BRANCH_SYNC_TIMEOUT, contextVariables } from '../constants';
-import { getGlobIgnorePatterns, readYML, getSyncIgnoreItems, shouldIgnorePath, getDefaultIgnorePatterns } from '../utils/common';
+import { getGlobIgnorePatterns, readYML, getSyncIgnoreItems, shouldIgnorePath, getDefaultIgnorePatterns, formatDatetime } from '../utils/common';
```
Comment (Basit-qc) [inline src/init/utils.ts:16]: Seems unused import

File: src/utils/config_utils.ts [hunk 11]
```diff
@@ -1,6 +1,7 @@
 import fs from 'fs';
 import { generateSettings } from "../settings";
 import { readYML } from "./common";
+import path from 'path';
```
Comment (Basit-qc) [inline src/utils/config_utils.ts:4]: Built-in imports should come on top

File: src/utils/tab_utils.ts [hunk 12]
```diff
@@ -0,0 +1,21 @@
+import fs from "fs";
+import { CodeSyncLogger } from '../logger';
+import { tabEventHandler } from "../events/tab_event_handler";
+import { formatDatetime } from "./common";
+
+export const removeTabFile = (filePath: string, funcName: string) => {
+	fs.unlink(filePath, err => {
```
Comment (Basit-qc) [inline src/utils/tab_utils.ts:7]: We should return if filePath does not exist, either i nthis funciton or before this function is called

File: tests/extension.test.js [hunk 13]
```diff
@@ -69,11 +74,33 @@ describe("Extension: activate", () => {
...
         global.IS_CODESYNC_TEST_MODE = true;
-
+        pathUtilsObj = new pathUtils(repoPath, DEFAULT_BRANCH);
+        tabsRepo = pathUtilsObj.getTabsRepo();
         fs.mkdirSync(repoPath, {recursive: true});
         fs.mkdirSync(baseRepoPath, {recursive: true});
+        fs.mkdirSync(tabsRepo, {recursive:true});
 
         createSystemDirectories();
+        mockTabs = [
```
Comment (Basit-qc) [inline tests/extension.test.js:84]: Seems a constant value, why not define outside of beforeEach

File: tests/test_init/utils.test.js [hunk 14]
```diff
@@ -254,6 +254,28 @@ describe("uploadRepo",  () => {
         "file_1.js": null,
     };
     expectedConfig[NESTED_PATH] = null;
+    const mockTabs = [
```
Comment (Basit-qc) [inline tests/test_init/utils.test.js:257]: If this is a constant response, why can't we define this in __mocks__?

File: tests/test_init/handler.test.js [hunk 15]
```diff
@@ -267,6 +267,32 @@ describe("initHandler: Syncing Branch", () => {
     });
 
     test("Should sync branch", async () => {
+        const mockTabs = [
```
Comment (Basit-qc) [inline tests/test_init/handler.test.js:270]: If this is a constant response, why can't we define this in mocks?

**Review by Basit-qc — CHANGES_REQUESTED** [review]

File: src/codesyncd/websocket/socket_events.ts (hunk 9 above)
Comment (saadberry) [reply src/codesyncd/websocket/socket_events.ts:153]: Good suggestion, thanks

File: src/codesyncd/websocket/socket_events.ts [hunk 16]
```diff
@@ -143,6 +193,23 @@ export class SocketEvents {
...
                 }
+            case EVENT_TYPES.TAB_PROCESSED:
+                switch (resp.status) {
+                    case HttpStatusCodes.OK:
+                        this.onTabProcessed(resp.file_name);
+                        return true;
+                    case HttpStatusCodes.PAYMENT_REQUIRED:
+                        this.onPaymentRequired();
+                        return true;
+                    case HttpStatusCodes.INVALID_USAGE:
```
[the remark this answers is not in the document]
Comment (saadberry) [reply src/codesyncd/websocket/socket_events.ts:204]: Sure, how about: "Unable to send tabs, status code <status_code> received from server" where status_code is either 400/403

File: src/codesyncd/validators/tab_validator.ts [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+
+	settings: any;
+	configJSON: any;
+	constructor() {
+		this.settings = generateSettings();
+        this.configJSON = readYML(this.settings.CONFIG_PATH);
+
+	}
+
+	validateRepo( repoPath: string ) {
```
Comment (Basit-qc) [inline src/codesyncd/validators/tab_validator.ts:17]: Formatting required here

File: src/codesyncd/websocket/socket_events.ts [hunk 18]
```diff
@@ -143,6 +191,25 @@ export class SocketEvents {
...
+                        this.onTabProcessed(resp.file_name);
+                        CodeSyncLogger.info(`Tab Event rejected by server, status_code=${HttpStatusCodes.INVALID_USAGE}`);
```
Comment (Basit-qc) [inline src/codesyncd/websocket/socket_events.ts:204]: Should be `CodeSyncLogger.error`

File: src/codesyncd/websocket/socket_events.ts [hunk 19]
```diff
@@ -143,6 +191,25 @@ export class SocketEvents {
...
+                        return true;
+                    case HttpStatusCodes.FORBIDDEN:
+                        this.onTabProcessed(resp.file_name);
+                        CodeSyncLogger.info(`Tab Event rejected by server, status_code=${HttpStatusCodes.FORBIDDEN}`);
```
Comment (Basit-qc) [inline src/codesyncd/websocket/socket_events.ts:208]: Should be `CodeSyncLogger.error`

File: src/utils/tab_utils.ts (hunk 12 above)
Comment (Basit-qc) [reply src/utils/tab_utils.ts:8]: I suggested to check if path exists or not, 
```if (!fs.existsSync(fielPath)) return;```

File: tests/test_codesyncd/buffer_handler.test.js [hunk 20]
```diff
@@ -632,4 +671,101 @@ describe("bufferHandler", () => {
...
+        };
+        const handled = await webSocketEvents.onMessage(msg);
+        expect(handled).toBe(true);
+        if (tabFiles.files.length) await waitFor(2);
+        // YML file should be deleted
+        assertTabFilesCount();
+        expect(fs.existsSync(path.join(tabsRepo, tabFiles.files[0]))).toBe(false);
+    });
+
+    test("SocketEvents: onMessage, Tab not sent (invalid usage)", async () => {
```
Comment (Basit-qc) [inline tests/test_codesyncd/buffer_handler.test.js:705]: Tab not sent -> Tab event was rejected

File: tests/test_codesyncd/buffer_handler.test.js [hunk 21]
```diff
@@ -632,4 +671,101 @@ describe("bufferHandler", () => {
...
+        };
+        const handled = await webSocketEvents.onMessage(msg);
+        expect(handled).toBe(true);
+        if (tabFiles.files.length) await waitFor(2);
+        // YML file should be deleted
+        assertTabFilesCount();
+        expect(fs.existsSync(path.join(tabsRepo, tabFiles.files[0]))).toBe(false);
+    });
+
+    test("SocketEvents: onMessage, Tab not sent (forbidden)", async () => {
```
Comment (Basit-qc) [inline tests/test_codesyncd/buffer_handler.test.js:727]: Tab not sent -> Tab event was rejected

File: tests/test_codesyncd/buffer_handler.test.js [hunk 22]
```diff
@@ -632,4 +671,101 @@ describe("bufferHandler", () => {
...
+        };
+        const handled = await webSocketEvents.onMessage(msg);
+        expect(handled).toBe(true);
+        if (tabFiles.files.length) await waitFor(2);
+        // YML file should be deleted
+        assertTabFilesCount();
+        expect(fs.existsSync(path.join(tabsRepo, tabFiles.files[0]))).toBe(false);
+    });
+
+    test("SocketEvents: onMessage, Tab not sent (payment required)", async () => {
```
Comment (Basit-qc) [inline tests/test_codesyncd/buffer_handler.test.js:749]: Tab event was rejected

File: tests/test_events/handler/tabEventHandler.test.js [hunk 23]
```diff
@@ -0,0 +1,180 @@
...
+    const pathUtilsObj = new pathUtils(repoPath, DEFAULT_BRANCH);
+    const tabsRepo = pathUtilsObj.getTabsRepo();
+    const newFilePath1 = path.join(repoPath, "file_1.js");
+    const newFilePath2 = path.join(repoPath, "directory", "documents", "new2.js");
+
+    beforeEach(() => {
+        jest.clearAllMocks();
+        // Create directories
+        fs.mkdirSync(repoPath, { recursive: true });
+        fs.mkdirSync(tabsRepo, { recursive: true });
```
Comment (Basit-qc) [inline tests/test_events/handler/tabEventHandler.test.js:40]: Why do we need this when we are already using `createSystemDirectories();`?

File: tests/test_events/handler/tabEventHandler.test.js [hunk 24]
```diff
@@ -0,0 +1,180 @@
...
+    const assertTabFilesCount = (tabsCount = 0) => {
+        let tabFiles = fs.readdirSync(tabsRepo);
+        expect(tabFiles).toHaveLength(tabsCount);
+        return true;
+    }
+
+    // No file should be created if no tab event occured
+    test("No tab", async () => {
+        configUtil.addRepo();
+        const handler = new bufferHandler(statusBarItem);
```
Comment (Basit-qc) [inline tests/test_events/handler/tabEventHandler.test.js:59]: Why do we have bufferHandler in the tabEventHandler tests?

File: tests/test_events/handler/tabEventHandler.test.js [hunk 25]
```diff
@@ -0,0 +1,180 @@
...
+        assertTabFilesCount();
+    });
+
+    // Skips case when repo exists
+    test("Should be skipped if repo does not exist",() => {
+        const handler = new tabEventHandler();
+        handler.handleTabChangeEvent();
+        // Verify no tab file should be generated
+        assertTabFilesCount()   
+        });
```
Comment (Basit-qc) [inline tests/test_events/handler/tabEventHandler.test.js:70]: FIx formatting

File: tests/extension.test.js [hunk 26]
```diff
@@ -69,9 +73,11 @@ describe("Extension: activate", () => {
         setWorkspaceFolders(repoPath);
         untildify.mockReturnValue(baseRepoPath);
         global.IS_CODESYNC_TEST_MODE = true;
-
+        pathUtilsObj = new pathUtils(repoPath, DEFAULT_BRANCH);
+        tabsRepo = pathUtilsObj.getTabsRepo();
         fs.mkdirSync(repoPath, {recursive: true});
         fs.mkdirSync(baseRepoPath, {recursive: true});
+        fs.mkdirSync(tabsRepo, {recursive:true});
```
Comment (Basit-qc) [inline tests/extension.test.js:80]: Why are we creating tabs directory separately when `createSystemDirectories();` should do that for us (used just below)

**Review by Basit-qc — CHANGES_REQUESTED** [review]

File: src/codesyncd/handlers/tab_handler.ts (hunk 4 above)
Comment (saadberry) [reply src/codesyncd/handlers/tab_handler.ts:11]: No we don't. Removing it

File: src/codesyncd/handlers/tabs_handler.ts [hunk 27]
```diff
@@ -0,0 +1,116 @@
...
+export class TabsHandler {
+
+    settings: any;
+
+    constructor() {}
+
+    static run(tabYmlFiles: ITabYML[]) {
+        const validTabs: ITabYML[] = [];
+        let tabsSize = 0;
+        if ( !tabYmlFiles ) return;
```
[the remark this answers is not in the document]
Comment (saadberry) [reply src/codesyncd/handlers/tabs_handler.ts:24]: You are correct. Thanks for pointing that out

File: src/utils/tab_utils.ts (hunk 12 above)
Comment (saadberry) [reply src/utils/tab_utils.ts:8]: Oh okay, thanks for clarifying

**Review by Basit-qc — APPROVED** [review]

**Pull request merged**